### DISCUSSION TITLE
Use Midi Controller to control the mixerboard

### DISCUSSION THREAD

**Discussion by henkdegroot** [jamulussoftware/jamulus#2220] [Technical Support & Help] [answered]
Hi,

I am trying to control the mixerboard in Jamulus using a Midi Controller. This seems to be working fine for the Fader, Pan and Solo.
However the Mute does not seem to pickup the midi command.

Currently I am sending Midi messages using a software/emulator (waiting for my actual midi controller to arrive).
I am using Windows with JACK.

Would appreciate if there is someone else who is using a midi controller can confirm this function (Mute) actually should work.
To be clear, I am referring to the MUTE option in the mixerboard, not the Mute Myself option.

Ideally I want to be able to quickly mute a group or individual persons while playing using my midi controller.
When I create a group in Jamulus and move a fader of one in the group, I see all are moving.
Next of course would be implementing the "group" solo/mute as these are not "group aware" yet.

Looking forward to tips/help.

**Comment by ann0see** [comment]
@ignotus666 should know more about this feature. I know it’s not 100% complete, but I‘d expect mute to work.

**Comment by ignotus666** [comment, marked as answer]
Yes, mute should work (it does for me). That's cool you're using it on Windows because officially it's not supposed to even be supported - I guess JACK is what let's you do that (note to self: if so this should be documented).

Anyway, maybe you could post your `--ctrlmidich` command and the notes your controller is using to see what might be going on.

**Reply by henkdegroot** [reply]
> guess JACK is what let's you do that

Yes, JACK brings this advantage to Jamulus in Windows. I am using the same ctrlmidich option as mentioned in the tips & tricks:

`--ctrlmidich '1;f0*8;p16*8;s32*8;m48*8'`

With the above I can use CC 1 to control the fader, CC 16 to control PAN, CC 32 to control Solo, and was expecting CC 48 for mute.

Changing the option to:

`--ctrlmidich '1;f0*8;p8*8;s16*8;m24*8'`

With the above I can use CC 1 to control the fader, CC 8 to control PAN, CC 16 to control Solo, and was expecting CC 24 for mute.

Here is a demo with the last setting active:
![MidiControl-Jamulus](https://user-images.githubusercontent.com/13550012/148605239-78029f86-9a6a-4a46-9ba0-7279a1b497f4.gif)

**Reply by henkdegroot** [reply]
Ha....found the issue! Think it requires an update to the documentation.....

`--ctrlmidich '1;f0*8;p8*8;s16*8;m24*8;'`

Note there is a semi-colon at the end....not sure if this is Windows only....

Here is the proof:
![MidiControl-Jamulus-working](https://user-images.githubusercontent.com/13550012/148606550-93554330-7e19-42aa-a7a4-090d309caed2.gif)

Silly....I can't mark my own answer as the answer :)

**Reply by henkdegroot** [reply]
@pcar75 who needs keyboard shortcuts when they can use a midi controller?
And note the prefix for the alias/name? The is automatically added when enabling the midi control option in Jamulus. Really clever solution.
Sorry...just realized now that this does not allow you to control the "Mute Myself" option. But that is one you should be able to control using a tool like autohotkey (at least on windows).

**Reply by ignotus666** [reply]
I just checked my config and without the semicolon it works for me on Linux. I'll check if it works _with_ it when I get a chance and update the doc. Anyway, glad you got it sorted :)

**Reply by henkdegroot** [reply]
Ahhh good old windows....it does not like the single quote being used when passing on the option. When I change to double quotes, it works on windows as well without the semi-colon at the end.

**Reply by ignotus666** [reply]
If double quotes work on all three OS's I wonder if it would be best to just mandate that instead of pointing out the exception of Windows.

**Reply by ignotus666** [reply]
Yep, works with double quotes on Linux. I suppose it would on macOS too.

**Reply by pcar75** [reply]
@henkdegroot 
> @pcar75 who needs keyboard shortcuts when they can use a midi controller?
> And note the prefix for the alias/name?

A midi controller would involve another piece of hardware, operated by hand(s) or foot (feet) would it not ?  Also, for my point of view, the inconvenience remains if one cannot access directly a channel.  Per pljones' example in thread #2213 , tabing to channel 114 would be a fastidious process and if it does not take hands away from the instrument (ex.: foot controller) , it would seriously take one's attention off the music .
My 2 cents ... :-)

> Sorry...just realized now that this does not allow you to control the "Mute Myself" option.

Note that Alt+M already controls the 'Mute Myself' button.

**Reply by henkdegroot** [reply]
@pcar75, this is the midi controller I am using:
![midi-controller-worlde](https://user-images.githubusercontent.com/13550012/148679134-ff48ef91-7d48-48b0-a6bb-1614bf58a1cf.png)

In fact I am using the same device from a different brand/supplier. This model seems to have been "copied" by several companies.
The Dutch company version has white led light buttons, not red (as the ones shown in the picture).

With this controller I can access 9 Jamulus clients in the mixerboard directly. No need for any tabbing, key press, etc. The button I am using to control the Mute. That is the downside for this controller, only 1 button. Korg has a similar controller which has 3 buttons, but that one is a little more expensive (not much). I selected my current model as it allows the define 4 banks (presets) which you can use to load different settings. I.e. one for when using Jamulus, one when using a DAW, etc.

**Reply by pcar75** [reply]
Nice , effectively a 'control surface' for DAW.  Perhaps it is possible to use 2 buttons to page right or left to show / access another set of 9 channels and reassign them the individual slider & controls ?

**Reply by henkdegroot** [reply]
> use 2 buttons to page right or left to show / access another set

This is already possible by switching using the Bank button. So with 4 banks, you can control 9 x 4 faders,pan,mute buttons.
The downside is that when switching the to another "set" of faders, the current position of the fader remains (of course as this device is manual operation only) and when you move that fader it starts at that position. Not sure how many you would need....

**Reply by pcar75** [reply]
I have not seen documentation indicating that 'Banks' could be interpreted by Jamulus client (ref.: https://jamulus.io/wiki/Tips-Tricks-More#using-ctrlmidich-for-midi-controllers ), to allow switching groups of channels.  How do you do that ?

[Edit] MIDI and the way it is implemented in Jamulus is puzzling to me !  Does anyone know if/how I can get Jamulus technical implementation doc.   And can anyone suggest a good book on MIDI ?

As stated by pljones, Jamulus can support 150 users in theory, which in real life would be overoptimistic (for now) I think, but I may be wrong. For realistic number of maximum channels to care for, public servers in these lists ...
- Any Genre 1, 2 and 3 and Genre Rock support an average maximum of 15-20 users
- Jazz, Classical/Folk and Barbershop/Choir support 30 or more (with wider variations in these genres).
- All this is evidently without regards to actual sound quality.
Ref: https://explorer.jamulus.io/

**Reply by ignotus666** [reply]
> I have not seen documentation indicating that 'Banks' could be interpreted by Jamulus client (ref.: https://jamulus.io/wiki/Tips-Tricks-More#using-ctrlmidich-for-midi-controllers ), to allow switching groups of channels. How do you do that ?

It's the external MIDI controller that switches banks; Jamulus just listens for individual CC numbers. Say you're in a Jamulus session with 16 musicians but your controller only has 8 faders (sending e.g. CC 0 - 7) - in order to control those last 8 channel strips in Jamulus you go up a bank in your controller so your 8 faders send CC 8 - 15.

AFAIK, you are limited to 128 CC numbers (0 - 127) for all parameters (fader, pan, mute and solo), as Jamulus only listens on one MIDI channel. So if you want to control all parameters via MIDI for all musicians, the maximum number of musicians you can control is 32. If we could assign a different MIDI channel to each parameter that would raise the limit to 128, and would also simplify the assignment of CC numbers for each parameter, or at least make it more intuitive - instead of offset (initial CC number) + consecutive CC numbers, which is fairly confusing, it could be initial CC number + channel. Going from this:
`--ctrlmidich '1;f0*8;p16*8;s32*8;m48*8'`
to something like this:
`--ctrlmidich 'f0*1;p0*2;s0*3;m0*4'`
But again, how hard this might be to implement is beyond me.

**Reply by henkdegroot** [reply]
> How do you do that ?

The Bank is an option for this specific device. Other devices may support this as well. It is a pre-set config for the device. So when I select Bank 1, all the button/faders/knobs have a specific function assigned. When I change to Bank 2, these assignments are changed to whatever is configured in the device for Bank 2.

Jamulus is currently able to "listen" to MIDI signals and can respond to specific events. Different types of MIDI events exists.
One of these events is known as Control Change (or CC in short) and they have a number in the range of 0 -> 127. Most events can also have a value assigned in the range of 0 -> 127. This value would like the fader position, or when using a single button the on (button pressed)/off (button released) value.
MIDI is transmitted using different channels (1 -> 16 in general). In Jamulus you can use channel 0 to indicate you don't care about the channel number and you want to "listen" to the information received from all channels.

In [Tips and Tricks Midi](https://jamulus.io/wiki/Tips-Tricks-More#using-ctrlmidich-for-midi-controllers) you can find a bit more about the command line option in Jamulus. I will use the basic here.

I start Jamulus using the command line option: `--ctrlmidich "1;f3*9"`

The 1 indicates I am telling Jamulus to listen to channel 1 (which happens to be the MIDI channel my device is using the transmit MIDI signals).
The f indicates I want to specify a CC for a Fader (fader in the mixer board). I start at 3 because my first fader will transmit CC number 3.
And because my device has 9 faders, the *9 is added. The Jamulus faders in the mixer board are now linked to the faders in my device. The first fader (Client 0) is listening to CC 3, the second fader to CC 4, the third to CC 5, etc.

When I now move fader 1 in my device, it sent out a MIDI message CC 3 + the value 0 (fader position at the bottom).
Jamulus receives this event and re-act by setting the fader for Client "0" (the most left) to 0.
If I move the fader up, another MIDI event message will be sent. Again with CC 3 + the value 64 (fader position in the center).
Jamulus receives this event and re-act by setting the fader for Client "0" (the most left) to 50%.

When I now move fader 5 in my device, it sent out a MIDI message CC 7 + the value 127 (fader position at the top).
Jamulus receives this event and re-act by setting the fader for Client "4" to 100%.

Buttons in MIDI work slightly different than fader. Most device can be set to have them as momentary or toggle switches.
Momentary would transmit two MIDI values, one for when you press the button (usually this sends the max 127 value) and one for when you release the button (usually this is the 0 value). So for this buttons Jamulus will receive two events. If you "link" this button to the Mute function, you will need to keep the button pressed for as long as you want to client to be muted (as when you release the button another signal is transmitted).

For Jamulus the toggle mode will better. First press of the button transmits value 127, second press of the button transmits value 0.
This buttons will also be configured to sent a specific CC number. As I am already using CC 3 -> 11 for the fader, I need to make sure my device sends out other CC events. Below each fader there is 1 button, which I configured the buttons to sent out CC 23 -> 31.

To make Jamulus aware of this, I have to update the command line option.

I use the command line option: `--ctrlmidich "1;f3*9;m23*9"`

So still having the faders here, but now added the m for the Mute button, indicate it starts with CC event 23 and again that I have 9.

When I now press the first button, it sent out MIDI message CC 23 + value 127. Because I enabled "toggle" mode nothing happens when I release the button.
Jamulus receives this event and re-act by enabling the Mute option for Client "0" (most left one).

When I now press the first button again, it sent out MIDI message CC 23 + value 0.
Jamulus receives this event and re-act by disabling the Mute option for Client "0" (most left one).

I hope this helps in understanding what is happening/how it works. I am planning on doing a KB article for this as well, but that will take some time.

**Reply by pcar75** [reply]
Thanks all for taking the time to answer my questions.

I was looking at this from a purely communication point of view but I now realize Jamulus MIDI implementation decisions were probably made from what existing MIDI control hardware implementation were available.
a) Most low to mid 'control surface' have either 1, 8/9 (mostly) or 16/17 physical channel strips.
b) MIDI CC channel number can only take values in 0-15 since the other fronting 4 bits indicate the message type (ex. CC message).
c) MIDI CC channel number is set to identify one specific device.

I keep my fingers crossed for instant quantum communication, and MIDI 2.0 or HUI or Mackie Control Protocols  and control automation/feedback  ;-)

MIDI :
A helpful reference (but dated) I found is http://www.music.mcgill.ca/~ich/classes/mumt306/StandardMIDIfileformat.html

Questions :
1. How are bank selection implemented in a 'control surface' and in Jamulus (i.e. bank1=mixer strip 1-8, bank2=mixer strip 9-16, etc.) ?  On bank2 selection, would Jamulus accordingly display channels 9 to 16 or more (my pc under Windows can show 18 channels at a time) ?

2. Can one initialize the Jamulus argument thus (the hw permitting) : 
`--ctrlmidich "1;f0*32;p32*32;m64*32;s96*32"`

Thanks in advance.

**Reply by henkdegroot** [reply]
Question 1:
Jamulus is not aware of the selected bank. Depending on the selected bank, the control surface will just sent out a different CC.
Here is what I have setup for my first bank:
![bank-scene-1](https://user-images.githubusercontent.com/13550012/149015988-97350000-a0a4-41cf-86f8-01fe035b98eb.PNG)

And here is what I have for the second bank:
![bank-scene-2](https://user-images.githubusercontent.com/13550012/149016043-d364e894-66a6-4415-a49b-e418dbe4835e.png)

On my pc I can easily get 50-60 channels displayed in one screen (using Compact mode and multiple rows).

Question 2:
Yes, the argument works fine.

**Reply by pcar75** [reply]
Bonjour again and thanks for info.
@henkdegroot 
> On my pc I can easily get 50-60 channels displayed in one screen (using Compact mode and multiple rows).

I cannot find how "Compact mode and multiple rows" can be activated in the Jamulus mixer window (albeit servers I have seen/used so far had less than 20 users concurrently. a view left-right slider appearing at the mixer bottom whenever needed) .   Is it a feature of the OS or display monitor you are using ? If so how ?
Thanks in advance.

**Reply by rdica** [reply]
In Jamulus 3.8.1, under My Profile tab;
Skin: Compact
Mixer Rows: 2 or more

I don't know if these options are available in Jamulus 3.6.2 (your JamulusOS thumb drive's version?)

**Reply by henkdegroot** [reply]
In Jamulus 3.6.2, you are limited to select 2 mixer channel rows max. In 3.8.0 support for max 8 rows has been added.
In 3.6.2 you can enable this in the Edit menu. 
In settings you can select the "Compact" skin, which will allow many more clients to be shown in mixer board.

**Comment by henkdegroot** [comment]
@ignotus666 
I received my midi controller today (a cheap one) but works really well. I can program it with unique CC numbers for each control.
This controller even has 4 banks, so I can select/switch between configs. Which I thought would be really nice to control all faders.
The downside is that when you switch, it does not/remember of course what level the fader was to previously.
I.e. I programed it so that the in bank 1 the first fader is CC 3 and in bank 2 it is CC 12. So the 10th client in jamulus can be controlled via bank 2 fader 1. However if the fader was set to like 75% while linked to client 1, it will start at 75% when you move it after the bank change.

So now thinking about buying 2 or 3 more of these controllers, so that I don't have to switch between banks and have a nice physical mixer desk for all connected clients. Would this work? I.e. using multiple controllers?

Another question, with the ctrlmidich option I can identify the channel to "listen" to. Should it be possible to use multiple devices each with it own midi channel? As I tried that but did not seem to work but I may have done/set the option incorrect.

**Reply by ignotus666** [reply]
Cool, that's how my midi controller works too, with two rows of 5 potentiometers that each have 9 banks, plus a few pots that have fixed values (my mic, input/output volume, etc.). What would be nice is a midi "pick up" feature: say the application's slider is at value CC75. It will only start moving when your controller's slider sends out that value, and from then on the two are "paired". Jack-mixer has this feature and I find it really useful to avoid e.g. blasting out something out at full volume when it should normally be at 20%. I have zero idea of how to implement this though...

**Reply by WolfganP** [reply]
https://xmmc.de/touchdaw/ is a configurable midi controller for Android, worth trying.
I paired it to jamulus on my client side and worked very well, but it was in a small band setup (so no idea in 10s of channels how well will perform with scrolling and the like)..

**Reply by ignotus666** [reply]
Yes, I tried something similar but not having a tablet it was on my 5.5" phone - not very practical. With a bigger screen it's an interesting option, but then again, why not just mirror your PC screen on your tablet and control Jamulus directly?

**Reply by pljones** [reply]
> why not just mirror your PC screen on your tablet and control Jamulus directly?

Bandwidth?

**Reply by thetome** [reply]
@henkdegroot 

> So now thinking about buying 2 or 3 more of these controllers, so that I don't have to switch between banks and have a nice physical mixer desk for all connected clients. Would this work? I.e. using multiple controllers?

Hi again,

got two Korg nanoKontrol working here with Jamulus now. Well, sort of... (BTW, not nanoKontrol 2, which is much more common and still sold. But I prefer the orginal nanoKontrol due to it's 9 instead of 8 channels and more buttons than the v2). 

I connected both of them via qJackctl's Graph captures successfully. Actually they both have an identical midi/CC setup, close to what you suggested in your examples, Henk. In scene/bank 1 both will move faders, knobs and pots as supposed and I can use both of them alternating or simultaneously. 

So far, so good. 

But switching banks in Jamuls via switching scenes within the controller doesn't work as expected on the first glance. I have the CC values set consecutive over the scenes, as you showed in the screenshots of your control editor.  Was that an assumption by you, or did you test this successfully (with an additional  controller emulator then, I guess) ? Here it leads to some funny behaviour: when e.g. scene/bank 2 is active, the first pan pot will toggle the mute button, solo button and slider will do nothing, and mute button will press/release solo...  In further scenes/banks it seem to get even more wild.
Now, I guess that can be solved with some thinking about the CC numbers. I'll go on with that and get back here when it's working ok.

But I just wanted to let you know, that in general it is possible to connect two (or even more) hardware controllers with Jamulus.

Best

**Reply by henkdegroot** [reply]
@thetome, at first I thought I had a similar issue. However it turned out that the software I used to program my control device did not correctly save the configuration. This resulted in some knobs/buttons/faders not being updated and therefore had duplicate CC numbers assigned. Maybe this happens with your device as well? I was able to spot this because the control device software can also read the current config.
I also found that some controllers do not change when switching banks in the control device. So be sure to select those controllers that do switch. I was able to use the setup with switching banks without issues. The only downside is the fader/knobs/buttons "memory" when switching. As it keeps the state/position after switching. Example, client 0 has the fader set to level 50, client 9 has the fader set to 70. Using the control device, I move the fader for client 0 to 20. Now I switch the bank, so the fader which previously controlled client 0 is now controlling client 9. Moving the fader "up" now makes the client 9 jump to 21 and then increases. For this reason I believe that switching banks is not really helping here.

**Reply by thetome** [reply]
@henkdegroot, thanks for the hint. A short test showed that my devices are switching banks properly. Problem was on user side (again), had to correct some parameters for the Jamulus shortcut, as well as a couple of CC values.

Working perfectly now, have 18 channels with vol, pan, solo, mute and mute myself direct at hand; plus bank switching.  Pretty cool :)

I'm used to the hopping faders when switching banks for some time now. Got a Zoom R16 as interface and controller with 8 passive faders in use with my DAW. I don't find it too annoying in most situations, it's worth the comfort plus for me.

Would Jamulus/Jack allow motorized faders at all?

**Reply by pljones** [reply]
There's no MIDI out currently.

**Reply by thetome** [reply]
I see, thank you. Anyway, such a great piece of software you guys are providing for free. 

I'll happily stay with the current setup, and am looking very forward to the next big jam, with many people and faders... 

![Jamulus Faders](https://user-images.githubusercontent.com/98780189/177041032-75f40639-6a1b-458c-ae10-03a8b68b7ae2.jpg)

**Comment by thetome** [comment]
Hi all and especially hello Henk,

I also have a Korg midi controller (nanoKontrol 1) that I would love to use with Jamulus in Windows. Its properly listed in device manager and works well in my DAW. I'm using the same Korg editor like you and can send and retrieve data to and from the device. Also I'm using loopMIDI in other context, if that helps.

I tried to follow your great example and tips here, but I can't get Jamulus to react to any movement from the controller so far. 
I guess it's Jack where I get lost, never used it before. I'm using Jack2, and my latest attempt was to start it via qjackctl.exe -S -X winmme , which should add midi support as I understand. But all midi entries that I can find then within qjackctl is system_midi, but I never see the Korg, or the 'MIDI controllers'-window from your animated screenshot. 

Could you enlighten me on how to setup Jack for recognizing my controller, and then send its midi to Jamulus, please?

Thank you very much, best regards
Tom

**Reply by henkdegroot** [reply]
> I can find then within qjackctl is system_midi,

That is correcct, it sounds like you are almost there. Just make sure you start Jamulus using the --ctrlmidich parameter and Jamulus should then show up in JACK with a midi input.
![image](https://user-images.githubusercontent.com/13550012/151965067-c3d5d460-16e1-41b3-96d1-3201a908f411.png)
Just connect the system midi to Jamulus to get the midi signal from the windows system into Jamulus.

When you restart Jamulus, the "midi" connection will be lost and you need to re-connect. I believe you can save this as a preference as well, I usually forget and then get surprised that the system no longer works.
Also important, make sure the midi device is connected before you start JACK.

**Reply by henkdegroot** [reply]
Also, you can check this document: [kb-midi-control-surface](https://jamulus.io/kb/2022/02/01/Midi-Mixerboard-Control.html)

**Comment by thetome** [comment]
Thank you so much much for your fast and useful help! Yesterday evening I tried to follow your advice, but suddenly I wasn't able to start the Jack server anymore for some reason (which worked the evening before, but then I tinkered around in many aspects of Jack..) Reinstalling Jack a couple of times didn't help. To get somewhere I'll try it on my wife's DAW and Jamulus PC later, it never had Jack installed so far.

And yes, your article is exactly what I am was dreaming of, should make it in the KB soon! Thanks again for all your effort.

I'll get back here once I got it somewhere...

**Reply by henkdegroot** [reply]
@thetome, I have also experienced some "odd" behaviour with JACK. Sometimes just clicking the Start button again will work. Another thing I have found that really helps is to go into the Setup -> Advanced tab and increase the Start Delay to something like 5 seconds. I think it works better this way as it has more time to start the software. And also make sure you have no other software using the audio hardware in "exclusive" mode.

**Reply by thetome** [reply]
@henkdegroot, The 5 sec delay obviously did the trick, server is running again (though it takes a couple of clicks on Start from time to time, as you said). I use Magix low latency ASIO driver as device for JACK, this is not in use by other SW. 

In JACK's Graph I now see the midi as supposed (I hope):

![grafik](https://user-images.githubusercontent.com/98780189/152170044-9c06ea23-3668-43fe-907c-7c6db7686b08.png)


But when I open Jamulus afterwards (using the same --ctrlmidich  parameters as given in your tutorial) nothing else shows up in JACK, it doesn't seem  to detect Jamulus (refreshing the Graph's view doesn't help). 

I think Jamulus is configured correctly regarding the midi, at least here's the fader with the leading number:

![grafik](https://user-images.githubusercontent.com/98780189/152168059-463e8a00-1b55-4c87-ab8e-03a2c1f4265c.png)


In your tutorial you state:

"Install Jamulus with JACK for Windows, accepting the default options but don't run Jamulus at the end."

Does this refer to: 

https://jamulus.io/de/kb/2020/11/24/Multiple-Audio-Interfaces.html  ?

I didn't do this so far, just have my Jamulus (3.8.1) with a plain installation. In Jamulus I have a Behringer UMC202HD as ASIO interface, this is probably wrong? Should it be the JackRouter instead?

**Reply by thetome** [reply]
Just switched Jamulus to Jackrouter, now I see it in the Graph. I'll dig into it.

**Reply by henkdegroot** [reply]
Note sure if that will be sufficient, i.e. does it show the "input midi"? If it does then it should work.
If not, then you need to download the Jamulus version was contains JACK integration (which does not use JackRouter). That version can be found here: https://github.com/jamulussoftware/jamulus/releases/download/r3_8_1/jamulus_3.8.1_win_jack.exe

**Reply by pljones** [reply]
> Just switched Jamulus to Jackrouter

That suggests to me you're using Jamulus for Windows ASIO, rather than Jamulus for Windows JACK?

**Reply by henkdegroot** [reply]
> Note sure if that will be sufficient, i.e. does it show the "input midi"?

I checked on my system and when using the Jackrouter (asio driver) and Jamulus with ASIO (instead of JACK), you don't get the "Input midi" in the JACK graph/dialog. So according to that, you have to make use of the Jamulus with jack version to make this work.
The "prefix" 0 is shown just because the ctrlmidich parameter was include when Jamulus was started. It does not mean that midi is active.

**Reply by thetome** [reply]
Oh my, thank you guys again for this clarification(s)!

Installed Jamulus with JACK and now I see the midi input too. Actually I haven't found the fitting system_midi capture yet, my system lists 10 of them. Which may be correct, I got many keyboards and controllers connected to my DAW PC. I tried to connect all of them simultaneously to Jamulus' input but didn't get slider :0 to move from any of all the hardware faders so far. Though, maybe a matter of midi offset (or similar) too,  I'll go and work through all the infos, especially your kb article, Henk. 

The journey continues, and thanks to all your support I guess I see land on the horizon now...

**Reply by henkdegroot** [reply]
If needed, I found this window tool very helpful: https://mountainutilities.eu/miditools
It can show you the MIDI messages being send by the device(s) and it can also be used to "send" MIDI data as if it is a device. Really nice for testing.

**Reply by pljones** [reply]
Oooh, nice - not quite as nice as MIDI-OX but at least it's got a 64bit version and is cross-platform.  Thanks!

**Reply by thetome** [reply]
Landed smoothly. A little try and error turned out it was capture_4, working fine now. 

Happy :)

Hartelijk bedankt, Henk! Thank you too, pljones!

**Comment by marcoboschi-1960** [comment]
hi all ..first of all thanks for your work and support :-)

i ve configured jamuluswithjack on win10 and korg nanocontrol2 ...and i successfully control sliders ,knobs , mute myself and solo buttons  (still not channel mute buttons ...dont understand why ... ) 

BUT...

 when i tried to use it in session  , i noticed that the number seems assigned based on initial positioning ...and that when someone leaves or join the session  the number is simply missed...so the slider became unusable ... also sorting produces (in my experience) no effects (ex. my channel remain number 4 and controlled by 4th slider  also if i select the check to move myself on first position )... i don't know if this is a problem in my installation or a normal and expected behaviour or maybe a bug ... but i suggest to try to associate the sliders in the visualized order (slide 1 on device controls the first displayed mixer Row ... at list as a selectable jamulus option...) 
thanks in advance for your response / comments

**Reply by pljones** [reply]
MIDI offset is based on the Jamulus channel number - it's nothing to do with how you have the GUI display sorted.  Unfortunately, despite it being requested, there's no "View -> By Channel Number" option.  I don't know how reliable the sort options are in general, to be honest.

**Reply by marcoboschi-1960** [reply]
Peter ... thanks for your kind reply ... I was mistakenly assuming it was possible to do some sort of offset channel translation by editing the message on the fly following some sort of client side mapping ...
thanks again for the wonderful work you are doing :-)

**Reply by marcoboschi-1960** [reply]
just to clarify ..it seems that the gui knows both the assigned channel number (is displayed under the slider) and obviously the order .of positioning... so i was assuming that the client could do the magic modifying the midi message on the fly ...

**Reply by marcoboschi-1960** [reply]
if inside the defined cc range ...

**Reply by pljones** [reply]
[Using ctrlmidich for MIDI controllers](https://jamulus.io/wiki/Tips-Tricks-More#using-ctrlmidich-for-midi-controllers)
I'm sure you've read the above...

The "channel number" mentioned there is the MIDI channel number - it doesn't affect the MIDI CC numbers in use, just which MIDI channel Jamulus will listen on for the MIDI CC messages.

You might also want to use your MIDI controller for other things...  Allocating a MIDI channel just for Jamulus is probably a good idea.

Now, there are five different things to control, only four of them are per Jamulus channel.  Those 128 MIDI CC messages need distributing.  So `--ctrlmidich '9;o0;f1*31;p32*31;s64*31;m96*31'` is about the most you can control: 31 Jamulus channels, plus server bypass.

Some servers may have more Jamulus channels than this.  So you could find a server having been full of 50 people later on only has 10 -- and none on those first 31 Jamulus channels.  Sorting solves nothing, of course - "4 x" would sort after all the "3x x" channels.

Getting back to the point here, then.

Are you saying that a channel that displays as in the range of the MIDI CC numbering you've assigned is not being correctly controlled?

You'll need to provide:
- the `--ctrlmidich` parameter you used
- the nanoKontrol2 settings and ideally a log of the values being sent (e.g. from a DAW's MIDI event trace) to see that it's working as expected
- a screen shot of a server you're connected to where it works as expected (if there are such cases), including the server genre and name
- a screen shot of a server you're connected to where it does not work as expected, including the server genre and name

Thanks!

**Comment by marcoboschi-1960** [comment]
Peter thanks again for reply ... and i don't want to waste yout time having not enought knoledge of the code .... 
i totally agree with you that the number of possible "controller numbers parameter" values for the 
"1011nnnn | 0ccccccc 0vvvvvvv  Control Change midi message" is limited and that we simply can't control more then, 32 mixing channel . on other hand i've to say that the majority of the session i've seen in real life has 10-20 people max ..and that usually the controllers used by home musician tipically hat 8-9 sliders lines ... 
having said that ... my guess was that we could tell the jamulus client to modifiy the midi message on the fly accondingly with the displayed order .. (maybe in the future a custom ordering could also be an option letting the user to decide which channels move to the left to  allow the control surface to control them) 

but as i said before .. i simply not had read and understood  the jamulus code before asking ... so i only assumed that was possible... seeing that the control channel is displyed on the mixer sliders (so jumulus knows it ) and jamulus for sure knows the config that i passed on the command line ...

I obviusly understand that also assuming that is possible , there is an effort for the coding and that the suggestion / implementation requests must be ordered and accepted based on number of affected users and effot for the development... 
.....so again thanks for the effort all you are doing to mantain and to evolve this great software :-)

**Reply by pljones** [reply]
> modifiy the midi message on the fly 

This just doesn't make sense.  The MIDI message stops being a MIDI message the moment it arrives in the message handler.

The CC number that the client gets via MIDI can only mean the server channel number (after working out which range it falls into and adjusting the server channel accordingly).  The client doesn't know anything about what the dialog box is displaying and the dialog box doesn't know anything about the MIDI messages.

Once the client determines the server channel from the controller number, it can update the server (and, dialog, if present).  The client doesn't even know what order the dialog is displaying the channels in.  The dialog still needs to use the server channel number to tell the client what to do, too.

**Reply by marcoboschi-1960** [reply]
Peter thank u very much for the patience and for the explanation...i understood how clean is the design of the software... 
... i really apologize ... for my bad english.... i had missed the meaning of the last sentence of the link you point me in the first reply ...
..... beeing the number in front of the name ...the by name ordering does the magic  ... simple and enough for make easy the controller 's phisical sliders usable...  :-)

**Reply by AndersGoran** [reply]
Picking up this old thread because it seems most closely related to what I'm trying to achieve. I would like to use a MIDI CC controller to control my own fader level - I don't care about the others, just my own. 

I play in a transatlantic setting with a painfully long latency. It is still tolerable for playing, but between songs when I _talk with the others_, as I hear my own voice inside my head immediately, it is unbearably annoying to hear an echo of what I just said 120 ms earlier. 

Thus I want to be able to quickly pull down my own level to zero, but not mute myself, because then the others will not hear what I say, right? And if I could control my __own fader__ with an expression pedal, I'd not have to put away my bow and reach for the mouse and fiddle with that. 

But maybe I'm trying to force the MIDI control feature to do something it can't? Suppose I set `--ctrlmidich "0;f100*1"`, i.e. listen on all MIDI channels, first fader CC is 100, allocate just one controller - that pedal is going to control channel ID 0, which can be anyone of the 3 or 4 (or 15) other people on the server, not necessarily me.

Changing the sort order of channels will not help, because the channel ID stays the same as long as I remain connected, so it doesn't matter if I'm the leftmost fader. I guess what I would need is for the faders on screen to correspond to the physical layout on my floor, and as long as my single pedal controlled the first fader, I'd be happy. But if I understand you correctly, @pljones, that's not how it works?

**Reply by rdica** [reply]
@AndersGoran Muting your channel will not stop others from hearing you. You just won't hear yourself through jamulus.
This is not the same as Mute Yourself button which **does** stop your client from sending audio data to the server, others would not hear you.

**Reply by henkdegroot** [reply]
@AndersGoran I believe there now is a function to have your own fader listed first. I did not check, but believe this means you can control your own fader.

**Reply by AndersGoran** [reply]
@rdica : thanks, but muting my channel via MIDI CC I suppose would still be tied to the channel ID which is fixed to me during the session, which is still a problem, because... 

@henkdegroot - I checked, it does not help. Here I am logged in as a **Listener** on a public server: 
<img width="289" alt="NotFirst" src="https://github.com/user-attachments/assets/7439d17d-29d3-4db8-9f62-b18c4ebebd4a">
I happened to get channel ID 1. Thus, since I now started Jamulus with `--ctrlmidich "0;f100*4"` and I have four controllers that send CC 100, 101, 102, 103 - a 4-port Arduino-based controller that I built myself. 

The first pedal with CC 100 controls the fader for "0:LowBot", the 2nd pedal with CC 101 controls my channel because I have ID 1, and the third pedal (CC 102) controls the last channel "2:".  

Now I switch to "me first", Cmd-W in the macOS client:
<img width="303" alt="First" src="https://github.com/user-attachments/assets/d86ef54c-a07f-43a3-a91b-89d355788fb1">
My ID is still 1 and the same pedal as before controls my fader. It makes no difference that the channel is displayed first. 

I would have to login to see which ID I get, log out, potentially reconfigure my controller (and change the `--ctrlmidich` argument) and hope nobody else logged in the meantime. It might work, but it's messy.

**Reply by henkdegroot** [reply]
Sounds like the my fader first does not help here indeed.

**Reply by AndersGoran** [reply]
I did a little hack just to see if I could pull it off: I have extended the code that parses `--ctrlmidich` to accept a "z" controller. This one works like a fader level, but is always hard-coded to adjust the level of `iMyChannelID`. 

This works fine, I now have a custom client that I can start with `--ctrlmidich "0;z100*1"` and CC 100 will always affect only my own fader.

**Reply by pljones** [reply]
Have you got a patch?  That sounds like it might be useful to others.

**Reply by AndersGoran** [reply]
I'm pretty lousy at pull requests and such stuff, but here's a somewhat cleaned-up patch: [new-ctrlmidich-z.zip](https://github.com/user-attachments/files/17230928/new-ctrlmidich-z.zip)

**Reply by pljones** [reply]
And here's the pull request: https://github.com/jamulussoftware/jamulus/pull/3394

**Comment by marcoboschi-1960** [comment]
sorry "control number" not channel :-)

**Comment by softins** [comment]
Just for anyone interested, Windows native MIDI support has recently been added to Jamulus, and will appear in the next release (probably 3.12.0). It will no longer be necessary to use the JACK version on Windows to get MIDI support.

**Reply by pljones** [reply]
Yep - we're hoping to cut an early dev release shortly, soon, as there's other stuff we'd like to get to more people to test -- so watch for an announcement 😄.

**Reply by thetome** [reply]
Great news, thank you very much.

**Reply by pljones** [reply]
The beta1 release should have this in.
https://github.com/jamulussoftware/jamulus/releases for "non release" releases, including this.

**Reply by rdica** [reply]
@pljones Thanks for the heads up on the beta release!
I didn't see an announcement regarding this, but I am beginning testing of servers and client now, thanks.
